scipy/stats/_stats_py.py

```diff
@@ -9018,64 +9029,136 @@ def confidence_interval(self, confidence_level=0.95):
 
         alternative = self._alternative
         p = self._p
-        x = np.sort(self._x)
-        n = len(x)
+        x = np.sort(self._x, axis=-1)
+        n = x.shape[-1]
         bd = stats.binom(n, p)
+        shape = p.shape
 
         if confidence_level <= 0 or confidence_level >= 1:
             message = "`confidence_level` must be a number between 0 and 1."
             raise ValueError(message)
 
-        low_index = np.nan
-        high_index = np.nan
-
-        if alternative == 'less':
+        if n == 0:
+            zeros = np.zeros(p.shape, dtype=x.dtype)
+            low, high = zeros, zeros
+        elif alternative == 'less':
             p = 1 - confidence_level
-            low = -np.inf
-            high_index = int(bd.isf(p))
-            high = x[high_index] if high_index < n else np.nan
+            low = np.full(shape, -np.inf)
+            high_index = bd.isf(p).astype(int)
+            valid_index = high_index < n
+            high_index[~valid_index] = 0
+            high = np.where(valid_index, np.take_along_axis(x, high_index), np.nan)
         elif alternative == 'greater':
             p = 1 - confidence_level
-            low_index = int(bd.ppf(p)) - 1
-            low = x[low_index] if low_index >= 0 else np.nan
-            high = np.inf
+            low_index = (bd.ppf(p)).astype(int) - 1
+            valid_index = low_index >= 0
+            low_index[~valid_index] = 0
+            low = np.where(valid_index, np.take_along_axis(x, low_index), np.nan)
+            high = np.full(shape, np.inf)
         elif alternative == 'two-sided':
             p = (1 - confidence_level) / 2
-            low_index = int(bd.ppf(p)) - 1
-            low = x[low_index] if low_index >= 0 else np.nan
-            high_index = int(bd.isf(p))
-            high = x[high_index] if high_index < n else np.nan
-
+            low_index = (bd.ppf(p)).astype(int) - 1
+            valid_index = low_index >= 0
+            low_index[~valid_index] = 0
+            low = np.where(valid_index, np.take_along_axis(x, low_index), np.nan)
+            high_index = bd.isf(p).astype(int)
+            valid_index = high_index < n
+            high_index[~valid_index] = 0
+            high = np.where(valid_index, np.take_along_axis(x, high_index), np.nan)
+
+        args = self._axis, self._axis_none, self._keepdims, self._ndim, self._nan_out
+        low = _quantile_test_postprocess(low, *args)
```

Comment:
This takes care of reducing away dimensions if `_keepdims=False`, adding dimensions back if `axis=None` and `keepdims=True` was passed, and NaNing-out values based on the nan-pattern of the input (or if the working axis. has zero length).